<launch>

  <arg name="ouster_ns" default="ouster" doc="Override the default namespace of all ouster nodes"/>
  <arg name="sensor_hostname" doc="hostname or IP in dotted decimal form of the sensor"/>
  <arg name="udp_dest" default=" " doc="hostname or IP where the sensor will send data packets"/>
  <arg name="lidar_port" default="0" doc="port to which the sensor should send lidar data"/>
  <arg name="imu_port" default="0" doc="port to which the sensor should send imu data"/>
  <arg name="udp_profile_lidar" default=" " doc="lidar packet profile; possible values: {
    LEGACY,
    RNG19_RFL8_SIG16_NIR16_DUAL,
    RNG19_RFL8_SIG16_NIR16,
    RNG15_RFL8_NIR8
    }"/>
  <arg name="lidar_mode" default=" " doc="resolution and rate; possible values: {
    512x10,
    512x20,
    1024x10,
    1024x20,
    2048x10,
    4096x5
    }"/>
  <arg name="timestamp_mode" default=" " doc="method used to timestamp measurements; possible values: {
    TIME_FROM_INTERNAL_OSC,
    TIME_FROM_SYNC_PULSE_IN,
    TIME_FROM_PTP_1588,
    TIME_FROM_ROS_TIME
    }"/>
  <arg name="metadata" default=" " doc="path to write metadata file when receiving sensor data"/>
  <arg name="viz" default="true" doc="whether to run a rviz"/>
  <arg name="rviz_config" default="$(find ouster_ros)/config/viz.rviz" doc="optional rviz config file"/>

  <arg name="tf_prefix" default=" " doc="namespace for tf transforms"/>
  <arg name="sensor_frame" default="os_sensor"
    doc="sets name of choice for the sensor_frame tf frame, value can not be empty"/>
  <arg name="lidar_frame" default="os_lidar"
    doc="sets name of choice for the os_lidar tf frame, value can not be empty"/>
  <arg name="imu_frame" default="os_imu"
    doc="sets name of choice for the os_imu tf frame, value can not be empty"/>
  <arg name="point_cloud_frame" default=" "
    doc="which frame to be used when publishing PointCloud2 or LaserScan messages.
    Choose between the value of sensor_frame or lidar_frame, leaving this value empty
    would set lidar_frame to be the frame used when publishing these messages."/>

  <arg name="no_bond" default="false"
    doc="request no bond setup when nodelets are created"/>
  <arg if="$(arg no_bond)" name="_no_bond" value="--no-bond"/>
  <arg unless="$(arg no_bond)" name="_no_bond" value=" "/>

  <arg name="proc_mask" default="IMG|PCL|IMU|SCAN" doc="
    use any combination of the 4 flags to enable or disable specific processors"/>

  <arg name="scan_ring" default="0" doc="
    use this parameter in conjunction with the SCAN flag
    and choose a value the range [0, sensor_beams_count)"/>

  <group ns="$(arg ouster_ns)">
    <node pkg="nodelet" type="nodelet" name="os_nodelet_mgr"
      output="screen" required="true"
      launch-prefix="bash -c 'sleep 2; $0 $@' "
      args="manager"/>
  </group>

  <group ns="$(arg ouster_ns)">
    <node pkg="nodelet" type="nodelet" name="os_driver"
      output="screen" required="true"
      args="load ouster_ros/OusterDriver os_nodelet_mgr $(arg _no_bond)">
      <param name="~/sensor_hostname" type="str" value="$(arg sensor_hostname)"/>
      <param name="~/udp_dest" type="str" value="$(arg udp_dest)"/>
      <param name="~/lidar_port" type="int" value="$(arg lidar_port)"/>
      <param name="~/imu_port" type="int" value="$(arg imu_port)"/>
      <param name="~/udp_profile_lidar" type="str" value="$(arg udp_profile_lidar)"/>
      <param name="~/lidar_mode" type="str" value="$(arg lidar_mode)"/>
      <param name="~/timestamp_mode" type="str" value="$(arg timestamp_mode)"/>
      <param name="~/metadata" type="str" value="$(arg metadata)"/>
      <param name="~/tf_prefix" value="$(arg tf_prefix)"/>
      <param name="~/sensor_frame" value="$(arg sensor_frame)"/>
      <param name="~/lidar_frame" value="$(arg lidar_frame)"/>
      <param name="~/imu_frame" value="$(arg imu_frame)"/>
      <param name="~/point_cloud_frame" value="$(arg point_cloud_frame)"/>
      <param name="~/proc_mask" value="$(arg proc_mask)"/>
      <param name="~/scan_ring" value="$(arg scan_ring)"/>
    </node>
  </group>

  <node if="$(arg viz)" pkg="rviz" name="rviz" type="rviz"
    output="screen" required="false" launch-prefix="bash -c 'sleep 5; $0 $@' "
    args="-d $(arg rviz_config)"/>

</launch>
